From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Sat, 25 Apr 2026 21:30:06 +1000
Subject: [PATCH] remove entity pushing


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 2d50cabf2ae898cbfb0396b46175e4b64ba4e206..2cba7a3ac14e08aeefebaa9927c36a8cb4c59638 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4104,58 +4104,6 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     protected void pushEntities() {
-        // Paper start - don't run getEntities if we're not going to use its result
-        if (!this.isPushable()) {
-            return;
-        }
-
-        net.minecraft.world.scores.Team team = this.getTeam();
-        if (team != null && team.getCollisionRule() == net.minecraft.world.scores.Team.CollisionRule.NEVER) {
-            return;
-        }
-
-        int i = ((ServerLevel) this.level()).getGameRules().get(GameRules.MAX_ENTITY_CRAMMING);
-        if (i <= 0 && this.level().paperConfig().collisions.maxEntityCollisions <= 0) {
-            return;
-        }
-        // Paper end - don't run getEntities if we're not going to use its result
-        // Leaf start - Only player pushable
-        final AABB box = this.getBoundingBox();
-        List<Entity> pushableEntities = org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled
-            ? getNearbyPushablePlayers(this, box, EntitySelector.pushableBy(this))
-            : this.level().getPushableEntities(this, box);
-        // Leaf end - Only player pushable
-        if (!pushableEntities.isEmpty()) {
-            if (this.level() instanceof ServerLevel serverLevel) {
-                // Paper - don't run getEntities if we're not going to use its result; moved up
-                if (i > 0 && pushableEntities.size() > i - 1 && this.random.nextInt(4) == 0) {
-                    int i1 = 0;
-
-                    for (Entity entity : pushableEntities) {
-                        if (!entity.isPassenger()) {
-                            i1++;
-                        }
-                    }
-
-                    if (i1 > i - 1) {
-                        this.hurtServer(serverLevel, this.damageSources().cramming(), 6.0F);
-                    }
-                }
-            }
-
-            // Paper start - Cap entity collisions
-            this.numCollisions = Math.max(0, this.numCollisions - this.level().paperConfig().collisions.maxEntityCollisions);
-            for (Entity entity1 : pushableEntities) {
-                if (this.numCollisions >= this.level().paperConfig().collisions.maxEntityCollisions) {
-                    break;
-                }
-
-                entity1.numCollisions++;
-                this.numCollisions++;
-                // Paper end - Cap entity collisions
-                this.doPush(entity1);
-            }
-        }
     }
 
     // Leaf start - Only player pushable
