From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Sun, 26 Apr 2026 02:12:52 +1000
Subject: [PATCH] retain player momentum after pearl teleport


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index fb282a076a977ade88c847ca3de5ff6d610a91be..6a51a9790509c7bfc6c847f164a0ce1e991ea34a 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -127,7 +127,16 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         // Store pre teleportation position as the teleport has been moved up.
                         final double preTeleportX = serverPlayer.getX(), preTeleportY = serverPlayer.getY(), preTeleportZ = serverPlayer.getZ();
                         final float preTeleportYRot = serverPlayer.getYRot(), preTeleportXRot = serverPlayer.getXRot();
-                        ServerPlayer serverPlayer1 = taskServerPlayer.teleport(new TeleportTransition(serverLevel, target, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
+
+                        final Vec3 velocity = EnderPearlMechanics.retainMomentum
+                            ? new Vec3(
+                                taskServerPlayer.getDeltaMovement().x,
+                                0.0,
+                                taskServerPlayer.getDeltaMovement().z
+                            )
+                            : Vec3.ZERO;
+
+                        ServerPlayer serverPlayer1 = taskServerPlayer.teleport(new TeleportTransition(serverLevel, target, velocity, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
                         if (serverPlayer1 == null) {
                             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
 
