From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Sun, 26 Apr 2026 02:12:52 +1000
Subject: [PATCH] retain player momentum after pearl teleport


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index e3f67101b7b6702eff18c3f64cf15a045fcb8b57..3b90e1b849010fccf2b2af64e3c6a7266b9229f5 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -29,6 +29,7 @@ import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import org.dreeam.leaf.config.modules.opt.PreventPearlGlitch;
+import org.dreeam.leaf.config.modules.opt.RetainPearlMomentum;
 import org.jspecify.annotations.Nullable;
 
 public class ThrownEnderpearl extends ThrowableItemProjectile {
@@ -121,7 +122,8 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         // Store pre teleportation position as the teleport has been moved up.
                         final double preTeleportX = serverPlayer.getX(), preTeleportY = serverPlayer.getY(), preTeleportZ = serverPlayer.getZ();
                         final float preTeleportYRot = serverPlayer.getYRot(), preTeleportXRot = serverPlayer.getXRot();
-                        ServerPlayer serverPlayer1 = taskServerPlayer.teleport(new TeleportTransition(serverLevel, target, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
+                        final Vec3 velocity = RetainPearlMomentum.enabled ? taskServerPlayer.getDeltaMovement() : Vec3.ZERO;
+                        ServerPlayer serverPlayer1 = taskServerPlayer.teleport(new TeleportTransition(serverLevel, target, velocity, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
                         if (serverPlayer1 == null) {
                             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
                             return;
