From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Sun, 26 Apr 2026 16:49:19 +1000
Subject: [PATCH] merge pearl mechanics


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 864e653ec3a4d7ca312c45d776fd28ac66f1512a..334bdbc3ef99a8a01fcf20573a8608c5b2f8e697 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -29,8 +29,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
-import org.dreeam.leaf.config.modules.opt.PreventPearlGlitch;
-import org.dreeam.leaf.config.modules.opt.RetainPearlMomentum;
+import org.dreeam.leaf.config.modules.opt.EnderPearlMechanics;
 import org.jspecify.annotations.Nullable;
 
 public class ThrownEnderpearl extends ThrowableItemProjectile {
@@ -110,7 +109,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
 
                 if (owner instanceof ServerPlayer serverPlayer) {
                     if (serverPlayer.connection.isAcceptingMessages()) {
-                        final Vec3 target = PreventPearlGlitch.enabled ? findPearlTarget(serverPlayer, serverLevel, hit) : hit;
+                        final Vec3 target = EnderPearlMechanics.preventGlitch ? findPearlTarget(serverPlayer, serverLevel, hit) : hit;
 
                         if (target == null) {
                             discard(EntityRemoveEvent.Cause.HIT);
@@ -124,7 +123,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         // Store pre teleportation position as the teleport has been moved up.
                         final double preTeleportX = serverPlayer.getX(), preTeleportY = serverPlayer.getY(), preTeleportZ = serverPlayer.getZ();
                         final float preTeleportYRot = serverPlayer.getYRot(), preTeleportXRot = serverPlayer.getXRot();
-                        final Vec3 velocity = RetainPearlMomentum.enabled ? taskServerPlayer.getDeltaMovement() : Vec3.ZERO;
+                        final Vec3 velocity = EnderPearlMechanics.retainMomentum ? taskServerPlayer.getDeltaMovement() : Vec3.ZERO;
                         ServerPlayer serverPlayer1 = taskServerPlayer.teleport(new TeleportTransition(serverLevel, target, velocity, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
                         if (serverPlayer1 == null) {
                             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
