From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Fri, 1 May 2026 00:18:33 +1000
Subject: [PATCH] preserve projectile ticking and visibility near border


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/MoonriseChunkLoadCounter.java b/ca/spottedleaf/moonrise/patches/chunk_system/MoonriseChunkLoadCounter.java
index c847d819b3de145d914f83cdbdfe1b80d36f908d..71e15c2ab3677b53b61a96ea8a0da45dab3918a3 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/MoonriseChunkLoadCounter.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/MoonriseChunkLoadCounter.java
@@ -31,7 +31,7 @@ public final class MoonriseChunkLoadCounter extends ChunkLoadCounter {
 
         for (int chunkX = chunkPos.x - chunkRadius; chunkX <= chunkPos.x + chunkRadius; ++chunkX) {
             for (int chunkZ = chunkPos.z - chunkRadius; chunkZ <= chunkPos.z + chunkRadius; ++chunkZ) {
-                if (!level.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+                if (!level.leaf$shouldAccessChunk(chunkX, chunkZ)) {
                     continue;
                 }
 
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
index c1d24cff2af534edfb9fcc45f3cf2b31eb04c6de..26b473fdee1272235cc0989ee81e617fa985cd71 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -12,6 +12,7 @@ import net.minecraft.util.AbortableIterationConsumer;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.projectile.Projectile;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.entity.EntityInLevelCallback;
@@ -419,7 +420,13 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         if (entity.isAlwaysTicking()) {
             return Visibility.TICKING;
         }
+
         final FullChunkStatus entityStatus = ((ChunkSystemEntity)entity).moonrise$getChunkStatus();
+
+        if (entity instanceof Projectile && entityStatus == FullChunkStatus.BLOCK_TICKING) {
+            return Visibility.TICKING;
+        }
+
         // Leaf start - optimize getEntityStatus
         return switch (entityStatus) {
             case INACCESSIBLE -> Visibility.HIDDEN;
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 7fa74bfbf808635c4b6a32898f3aeb0851e4811f..b6f5bbdf2b421d0c2a4c20abb892113440707ab3 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -776,10 +776,10 @@ public final class RegionizedPlayerChunkLoader {
 
         private boolean shouldTrackChunk(final int chunkX, final int chunkZ) {
             return !trackByWorldBorder || (
-                chunkX >= trackMinChunkX &&
-                chunkX <= trackMaxChunkX &&
-                chunkZ >= trackMinChunkZ &&
-                chunkZ <= trackMaxChunkZ
+                chunkX >= trackMinChunkX - 1 &&
+                chunkX <= trackMaxChunkX + 1 &&
+                chunkZ >= trackMinChunkZ - 1 &&
+                chunkZ <= trackMaxChunkZ + 1
             );
         }
 
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index b05eddf3e099e37fcf614827e972bf3e958c0431..7b3bc0a020db6006b8ce728bcf07576c5b293476 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -397,7 +397,7 @@ public final class ChunkHolderManager {
 
                 if (current == null) {
                     // must create
-                    if (!ChunkHolderManager.this.world.leaf$shouldTrackChunk(CoordinateUtils.getChunkX(key), CoordinateUtils.getChunkZ(key))) {
+                    if (!ChunkHolderManager.this.world.leaf$shouldAccessChunk(CoordinateUtils.getChunkX(key), CoordinateUtils.getChunkZ(key))) {
                         iterator.remove();
 
                         continue;
@@ -629,7 +629,8 @@ public final class ChunkHolderManager {
 
         final int chunkX = CoordinateUtils.getChunkX(chunk);
         final int chunkZ = CoordinateUtils.getChunkZ(chunk);
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             return false;
         }
 
@@ -981,7 +982,8 @@ public final class ChunkHolderManager {
     private NewChunkHolder createChunkHolder(final long position) {
         final int chunkX = CoordinateUtils.getChunkX(position);
         final int chunkZ = CoordinateUtils.getChunkZ(position);
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             throw new IllegalStateException("Cannot create chunk holder outside the world border: (" + chunkX + "," + chunkZ + ")");
         }
 
@@ -1002,7 +1004,8 @@ public final class ChunkHolderManager {
     private NewChunkHolder getOrCreateChunkHolder(final long position) {
         final int chunkX = CoordinateUtils.getChunkX(position);
         final int chunkZ = CoordinateUtils.getChunkZ(position);
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             throw new IllegalStateException("Cannot create chunk holder outside the world border: (" + chunkX + "," + chunkZ + ")");
         }
 
@@ -1034,7 +1037,7 @@ public final class ChunkHolderManager {
             TickThread.ensureTickThread(this.world, chunkX, chunkZ, "Cannot create entity chunk off-main");
         }
 
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             throw new IllegalStateException("Cannot create entity chunk outside the world border: (" + chunkX + "," + chunkZ + ")");
         }
 
@@ -1122,7 +1125,7 @@ public final class ChunkHolderManager {
             TickThread.ensureTickThread(this.world, chunkX, chunkZ, "Cannot create poi chunk off-main");
         }
 
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             throw new IllegalStateException("Cannot create poi chunk outside the world border: (" + chunkX + "," + chunkZ + ")");
         }
 
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index b784c5ae99b60240138e4c9d46f5e69cc7143c41..f5dbe92f16a7b8026bc492658d3fb0a66e05f065 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -447,7 +447,7 @@ public final class ChunkTaskScheduler {
             throw new IllegalArgumentException("Cannot wait for INACCESSIBLE status");
         }
 
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             if (onComplete != null) {
                 try {
                     onComplete.accept(null);
@@ -556,9 +556,8 @@ public final class ChunkTaskScheduler {
     }
 
     // only appropriate to use with syncLoadNonFull
-    public boolean beginChunkLoadForNonFullSync(final int chunkX, final int chunkZ, final ChunkStatus toStatus,
-                                                final Priority priority) {
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+    public boolean beginChunkLoadForNonFullSync(final int chunkX, final int chunkZ, final ChunkStatus toStatus, final Priority priority) {
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             return false;
         }
 
@@ -611,7 +610,7 @@ public final class ChunkTaskScheduler {
             return this.world.getChunkSource().getChunk(chunkX, chunkZ, status, true);
         }
 
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             throw new IllegalStateException(
                 "Cannot sync load chunk outside the world border: (" + chunkX + "," + chunkZ + ") status: " + status
             );
@@ -674,7 +673,7 @@ public final class ChunkTaskScheduler {
             return;
         }
 
-        if (!world.leaf$shouldTrackChunk(chunkX, chunkZ)) {
+        if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
             if (onComplete != null) {
                 try {
                     onComplete.accept(null);
@@ -775,7 +774,7 @@ public final class ChunkTaskScheduler {
     }
 
     private boolean shouldUseBorderChunk(final int chunkX, final int chunkZ) {
-        return !world.leaf$shouldTrackChunk(chunkX, chunkZ);
+        return !world.leaf$shouldAccessChunk(chunkX, chunkZ);
     }
 
     ChunkProgressionTask schedule(final int chunkX, final int chunkZ, final ChunkStatus targetStatus, final NewChunkHolder chunkHolder,
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
index c5bd444bf570da2a363613ddf651d7d2f46d2c35..916c3688e1c95d20f11c8b092dcc5e1e4ec0615a 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
@@ -1178,7 +1178,7 @@ public final class NewChunkHolder {
                     final NewChunkHolder holder = (dx | dz) == 0 ? this : this.scheduler.chunkHolderManager.getChunkHolder(dx + this.chunkX, dz + this.chunkZ);
                     if (loaded) {
                         if (holder == null) {
-                            if (!world.leaf$shouldTrackChunk(dx + chunkX, dz + chunkZ)) {
+                            if (!world.leaf$shouldAccessChunk(dx + chunkX, dz + chunkZ)) {
                                 continue;
                             }
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 098b03bf088a9e214fd65322225f777c12985fb8..3012513a78a7c20708df36442e427c79ed6fce3b 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2602,6 +2602,33 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         return maxBlockX >= (int) Math.floor(worldBorder.getMinX()) && minBlockX <= (int) Math.ceil(worldBorder.getMaxX()) - 1
             && maxBlockZ >= (int) Math.floor(worldBorder.getMinZ()) && minBlockZ <= (int) Math.ceil(worldBorder.getMaxZ()) - 1;
     }
+
+    public final boolean leaf$shouldAccessChunk(final int chunkX, final int chunkZ) {
+        if (!galeConfig().smallOptimizations.loadChunks.preventTrackingOutsideWorldBorder) {
+            return true;
+        }
+
+        if (leaf$shouldTrackChunk(chunkX, chunkZ)) {
+            return true;
+        }
+
+        for (int dz = -1; dz <= 1; ++dz) {
+            for (int dx = -1; dx <= 1; ++dx) {
+                if ((dx | dz) == 0) {
+                    continue;
+                }
+
+                if (!leaf$shouldTrackChunk(dx + chunkX, dz + chunkZ)) {
+                    continue;
+                }
+
+                return true;
+            }
+        }
+
+        return false;
+    }
+
     // Leaf end - Do not track chunks outside the world border
 
     @Override
@@ -2966,7 +2993,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             //this.entityManager.processPendingLoads(); // Paper - rewrite chunk system
 
             for (ChunkPos chunkPos1 : list) {
-                if (!leaf$shouldTrackChunk(chunkPos1.x, chunkPos1.z)) {
+                if (!leaf$shouldAccessChunk(chunkPos1.x, chunkPos1.z)) {
                     continue;
                 }
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 2cfbe1537c2f8a405fe798df8adbb3bcda326977..2bb74d0ecb71ca9c8000972c5a3cccb6fde058ed 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -4,7 +4,6 @@ import com.google.common.collect.Lists;
 import com.mojang.serialization.Codec;
 import java.io.IOException;
 import java.util.Collection;
-import java.util.Iterator;
 import java.util.List;
 import java.util.UUID;
 import java.util.function.Consumer;
@@ -12,7 +11,6 @@ import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
-import net.minecraft.ReportedException;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
@@ -1063,19 +1061,39 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         // CraftBukkit end
         if (this.isOutsideBuildHeight(pos)) {
             return Blocks.VOID_AIR.defaultBlockState();
-        } else {
-            ChunkAccess chunk = this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4);
+        }
+
+        final int chunkX = pos.getX() >> 4;
+        final int chunkZ = pos.getZ() >> 4;
 
-            return chunk == null ? null : chunk.getBlockState(pos);
+        ChunkAccess chunkAccess = getChunkIfLoadedImmediately(chunkX, chunkZ);
+
+        if (chunkAccess == null &&
+            this instanceof final ServerLevel serverLevel &&
+            serverLevel.leaf$shouldAccessChunk(chunkX, chunkZ)
+        ) {
+            chunkAccess = this.getChunk(chunkX, chunkZ, ChunkStatus.FULL, false);
         }
+
+        return chunkAccess == null ? null : chunkAccess.getBlockState(pos);
     }
 
     @Override
     @Nullable
     public final FluidState getFluidIfLoaded(BlockPos pos) {
-        ChunkAccess chunk = this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4);
+        final int chunkX = pos.getX() >> 4;
+        final int chunkZ = pos.getZ() >> 4;
+
+        ChunkAccess chunkAccess = this.getChunkIfLoadedImmediately(chunkX, chunkZ);
 
-        return chunk == null ? null : chunk.getFluidState(pos);
+        if (chunkAccess == null &&
+            this instanceof final ServerLevel serverLevel &&
+            serverLevel.leaf$shouldAccessChunk(chunkX, chunkZ)
+        ) {
+            chunkAccess = this.getChunk(chunkX, chunkZ, ChunkStatus.FULL, false);
+        }
+
+        return chunkAccess == null ? null : chunkAccess.getFluidState(pos);
     }
 
     @Override
@@ -1770,6 +1788,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     @Override
     public @Nullable BlockGetter getChunkForCollisions(int chunkX, int chunkZ) {
+        if (this instanceof final ServerLevel serverLevel && !serverLevel.leaf$shouldAccessChunk(chunkX, chunkZ)) {
+            return null;
+        }
+
         return this.getChunk(chunkX, chunkZ, ChunkStatus.FULL, false);
     }
 
@@ -2194,7 +2216,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return this.moonrise$getEntityLookup().getEntityCount(); // Paper - rewrite chunk system
     }
     // Paper end - allow patching this logic
-    
+
     // Purpur start - Add allow water in end world option
     public boolean isNether() {
         return getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER;
