From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Mon, 4 May 2026 21:06:00 +1000
Subject: [PATCH] fix legacy tracker initial visibility


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index aaee85544696e822bbc08e0264ed3fec86731a14..4923eb3aa601b3858e1febfb55a0e3fe93907317 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1,7 +1,7 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Lists;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.MapCodec;
@@ -1090,31 +1090,22 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         // Paper end - optimise entity tracker
         // Paper - rewrite chunk system
 
-        List<ServerPlayer> list = Lists.newArrayList();
-        List<ServerPlayer> list1 = this.level.players();
-
         for (ChunkMap.TrackedEntity trackedEntity : this.entityMap.values()) {
-            SectionPos sectionPos = trackedEntity.lastSectionPos;
-            SectionPos sectionPos1 = SectionPos.of(trackedEntity.entity);
-            boolean flag = !Objects.equals(sectionPos, sectionPos1);
-            if (flag) {
-                trackedEntity.updatePlayers(list1);
-                Entity entity = trackedEntity.entity;
-                if (entity instanceof ServerPlayer) {
-                    list.add((ServerPlayer)entity);
-                }
+            final Entity entityTracked = trackedEntity.entity;
+            final SectionPos lastPos = trackedEntity.lastSectionPos;
+            final SectionPos currentPos = SectionPos.of(entityTracked);
+            final boolean isDifferent = !Objects.equals(lastPos, currentPos);
 
-                trackedEntity.lastSectionPos = sectionPos1;
-            }
+            trackedEntity.moonrise$tick(
+                ((ChunkSystemEntity) entityTracked).moonrise$getChunkData().nearbyPlayers
+            );
 
-            if (flag || trackedEntity.entity.needsSync || this.distanceManager.inEntityTickingRange(sectionPos1.chunk().toLong())) {
-                trackedEntity.serverEntity.sendChanges();
+            if (isDifferent) {
+                trackedEntity.lastSectionPos = currentPos;
             }
-        }
 
-        if (!list.isEmpty()) {
-            for (ChunkMap.TrackedEntity trackedEntity : this.entityMap.values()) {
-                trackedEntity.updatePlayers(list);
+            if (isDifferent || entityTracked.needsSync || this.distanceManager.inEntityTickingRange(currentPos.chunk().toLong())) {
+                trackedEntity.serverEntity.sendChanges();
             }
         }
     }
