From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: plantball <mythicalplanetdude@gmail.com>
Date: Sat, 13 Jun 2026 00:56:19 +1000
Subject: [PATCH] harden world border chunk access


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index 7824d3277008830a95d65d2382788942feb0173f..132f6452dc4a4bbacfb5e604991c14cb791a8d44 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -1103,7 +1103,12 @@ public final class ChunkHolderManager {
         }
 
         if (!world.leaf$shouldAccessChunk(chunkX, chunkZ)) {
-            throw new IllegalStateException("Cannot create entity chunk outside the world border: (" + chunkX + "," + chunkZ + ")");
+            LOGGER.warn(
+                "Allowing entity chunk creation outside tracked bounds in world {} at ({}, {})",
+                world.getWorld().getName(),
+                chunkX,
+                chunkZ
+            );
         }
 
         ChunkEntitySlices ret;
